File: logisim/app/src/main/java/logisim/gui/generic/CanvasPane.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.gui.generic;

import java.awt.Component;
import java.awt.Dimension;
import java.awt.Rectangle;
import java.awt.event.ComponentEvent;
import java.awt.event.ComponentListener;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import javax.swing.JScrollPane;
import javax.swing.SwingConstants;

import logisim.util.MacCompatibility;

public class CanvasPane extends JScrollPane {
	private class Listener implements ComponentListener, PropertyChangeListener {

		//
		// ComponentListener methods
		//
		public void componentResized(ComponentEvent e) {
			contents.recomputeSize();
		}

		public void componentMoved(ComponentEvent e) {
		}

		public void componentShown(ComponentEvent e) {
		}

		public void componentHidden(ComponentEvent e) {
		}

		public void propertyChange(PropertyChangeEvent e) {
			String prop = e.getPropertyName();
			if (prop.equals(ZoomModel.ZOOM)) {
				double oldZoom = (Double) e.getOldValue();
				Rectangle r = getViewport().getViewRect();
				double cx = (r.x + (double) r.width / 2) / oldZoom;
				double cy = (r.y + (double) r.height / 2) / oldZoom;

				double newZoom = (Double) e.getNewValue();
				contents.recomputeSize();
				r = getViewport().getViewRect();
				int hv = (int) (cx * newZoom) - r.width / 2;
				int vv = (int) (cy * newZoom) - r.height / 2;
				getHorizontalScrollBar().setValue(hv);
				getVerticalScrollBar().setValue(vv);
			}
		}
	}

	private CanvasPaneContents contents;
	private Listener listener;
	private ZoomModel zoomModel;

	public CanvasPane(CanvasPaneContents contents) {
		super((Component) contents);
		this.contents = contents;
		listener = new Listener();
		zoomModel = null;
		if (MacCompatibility.mrjVersion >= 0.0) {
			setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
			setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
		}

		addComponentListener(listener);
		contents.setCanvasPane(this);
	}

	public void setZoomModel(ZoomModel model) {
		ZoomModel oldModel = zoomModel;
		if (oldModel != null) oldModel.removePropertyChangeListener(ZoomModel.ZOOM, listener);
		zoomModel = model;
		if (model != null) model.addPropertyChangeListener(ZoomModel.ZOOM, listener);
	}

	public double getZoomFactor() {
		ZoomModel model = zoomModel;
		return model == null ? 1.0 : model.getZoomFactor();
	}

	public Dimension getViewportSize() {
		Dimension size = new Dimension();
		getViewport().getSize(size);
		return size;
	}

	public int supportScrollableBlockIncrement(Rectangle visibleRect, int direction) {
		int unit = supportScrollableUnitIncrement();
		if (direction == SwingConstants.VERTICAL) return visibleRect.height / unit * unit;
		else return visibleRect.width / unit * unit;
	}

	public int supportScrollableUnitIncrement() {
		double zoom = getZoomFactor();
		return (int) Math.round(10 * zoom);
	}

	public Dimension supportPreferredSize(int width, int height) {
		double zoom = getZoomFactor();
		if (zoom != 1.0) {
			width = (int) Math.ceil(width * zoom);
			height = (int) Math.ceil(height * zoom);
		}
		Dimension minSize = getViewportSize();
		if (minSize.width > width)
			width = minSize.width;
		if (minSize.height > height)
			height = minSize.height;
		return new Dimension(width, height);
	}
}

File: LogisimPlus/LogisimPlus/logisim/gui/generic/CanvasPane.cs
﻿// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.gui.generic
{


	using MacCompatibility = logisim.util.MacCompatibility;

	public class CanvasPane : JScrollPane
	{
		private class Listener : ComponentListener, PropertyChangeListener
		{
			private readonly CanvasPane outerInstance;

			public Listener(CanvasPane outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			//
			// ComponentListener methods
			//
			public virtual void componentResized(ComponentEvent e)
			{
				outerInstance.contents.recomputeSize();
			}

			public virtual void componentMoved(ComponentEvent e)
			{
			}

			public virtual void componentShown(ComponentEvent e)
			{
			}

			public virtual void componentHidden(ComponentEvent e)
			{
			}

			public virtual void propertyChange(PropertyChangeEvent e)
			{
				string prop = e.getPropertyName();
				if (prop.Equals(ZoomModel.ZOOM))
				{
					double oldZoom = ((double?) e.getOldValue()).Value;
					Rectangle r = getViewport().getViewRect();
					double cx = (r.X + r.Width / 2) / oldZoom;
					double cy = (r.Y + r.Height / 2) / oldZoom;

					double newZoom = ((double?) e.getNewValue()).Value;
					outerInstance.contents.recomputeSize();
					r = getViewport().getViewRect();
					int hv = (int)(cx * newZoom) - r.Width / 2;
					int vv = (int)(cy * newZoom) - r.Height / 2;
					getHorizontalScrollBar().setValue(hv);
					getVerticalScrollBar().setValue(vv);
				}
			}
		}

		private CanvasPaneContents contents;
		private Listener listener;
		private ZoomModel zoomModel;

		public CanvasPane(CanvasPaneContents contents) : base((Component) contents)
		{
			this.contents = contents;
			this.listener = new Listener(this);
			this.zoomModel = null;
			if (MacCompatibility.mrjVersion >= 0.0)
			{
				setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
				setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
			}

			addComponentListener(listener);
			contents.CanvasPane = this;
		}

		public virtual ZoomModel ZoomModel
		{
			set
			{
				ZoomModel oldModel = zoomModel;
				if (oldModel != null)
				{
					oldModel.removePropertyChangeListener(ZoomModel.ZOOM, listener);
				}
				zoomModel = value;
				if (value != null)
				{
					value.addPropertyChangeListener(ZoomModel.ZOOM, listener);
				}
			}
		}

		public virtual double ZoomFactor
		{
			get
			{
				ZoomModel model = zoomModel;
				return model == null ? 1.0 : model.ZoomFactor;
			}
		}

		public virtual Size ViewportSize
		{
			get
			{
				Size size = new Size();
				getViewport().getSize(size);
				return size;
			}
		}

		public virtual int supportScrollableBlockIncrement(Rectangle visibleRect, int orientation, int direction)
		{
			int unit = supportScrollableUnitIncrement(visibleRect, orientation, direction);
			if (direction == SwingConstants.VERTICAL)
			{
				return visibleRect.height / unit * unit;
			}
			else
			{
				return visibleRect.width / unit * unit;
			}
		}

		public virtual int supportScrollableUnitIncrement(Rectangle visibleRect, int orientation, int direction)
		{
			double zoom = ZoomFactor;
			return (int) (long)Math.Round(10 * zoom, MidpointRounding.AwayFromZero);
		}

		public virtual Size supportPreferredSize(int width, int height)
		{
			double zoom = ZoomFactor;
			if (zoom != 1.0)
			{
				width = (int) Math.Ceiling(width * zoom);
				height = (int) Math.Ceiling(height * zoom);
			}
			Size minSize = ViewportSize;
			if (minsize.Width > width)
			{
				width = minsize.Width;
			}
			if (minsize.Height > height)
			{
				height = minsize.Height;
			}
			return new Size(width, height);
		}
	}

}
